text/0626-invoke-helper.md

```diff
@@ -0,0 +1,564 @@
+        }
+
+        return this.data;
+      }
+
+      willDestroy() {
+        cancelFetch(this.promise);
+      }
+    }
+    ```
+    ```hbs
+    <!-- app/components/search.hbs -->
+    {{doc-title 'Search Result for "{{@query}}"'}}
+
+    {{#let (fetch-data "www.example.com/search?query={{@query}}") as |data|}}
+      ...
+    {{/let}}
+    ```
+
+Out of these options, helpers are the closest to what we want - they produce
+computed values directly without a template, and with the recent addition of
+effect helpers they can be used side-effect to accomplish tasks like setting the
+document title. The only downside is that they can only be invoked in templates,
+so they require you to design your components around using them in templates
+only. This can be difficult to do in many cases, where the data wants to be
+accessed to create derived state for instance.
+
+This RFC proposes adding a way to create helpers within JavaScript directly,
+extending the reactive model in a way that allows users to extract common
+reactive code and patterns, and reuse them transparently. This will make helpers
+the new reactive atom of the system - the reactive equivalent of a "function" in
+our incremental model. Like components, they have a lifecycle, and can update
+over time. Unlike components, they can exist nested in JavaScript classes _and_
+in templates, and they can produce any type of value, making them much more
+flexible.
+
+```js
+// app/components/search.js
+import Component from '@ember/component';
+
+export default class Search extends Component {
+  data = invokeHelper(this, FetchData, () => {
+    return {
+      positional: [`www.example.com/search?query=${this.query}`],
+    };
+  });
+
+  constructor() {
+    super(...arguments);
+
+    invokeHelper(this, DocTitle, () => {
+      return {
+        positional: [`Search Result for "${this.query}"`],
+      };
+    });
+  }
+}
+```
+
+Note how the concerns are completely separated in this version of the component.
+The polling logic is self contained, and separated from the data fetching logic.
+Both sets of logic are able to contain their lifecycles, updating based on
+changes to tracked state, and tearing down when the program destroys them. In
+the future, convenience APIs can be added to make invoking them easier to read:
+
+```js
+export default class Search extends Component {
+  @use data = fetchData(() => `www.example.com/search?query=${this.query}`);
+
+  constructor() {
+    super(...arguments);
+
+    use(this, docTitle(() => `Search Result for "${this.query}"`));
+  }
+}
+```
+
+## Detailed design
+
+This RFC proposes adding the `invokeHelper` function, imported from
+`@ember/helper`. The function will have the following interface (using
+TypeScript types for brevity and clarity):
+
+```ts
+interface TemplateArgs {
+  positional?: unknown[],
+  named?: Record<string, unknown>
+}
+
+type HelperDefinition = object;
+
+interface Helper {
```

Comment:
I still think this is a large issue. This is **not** a "helper", this is like a "helper value wrapper thingy". 🤔